ipaserver/install/ipa_migrate.py

```diff
@@ -0,0 +1,1821 @@
+    'schema_oc_skipped': 0,
+    'schema_processed': 0,
+    # Config
+    'config_processed': 0,
+    'config_migrated': 0,
+    # general
+    'conflicts': 0,  # LDIF log entries
+    'ignored_errors': 0,  # force option
+    # db
+    'reset_range': 0,
+    'total_db_entries': 0,
+    'total_db_migrated': 0,
+}
+
+
+#
+# Generic helper functions
+#
+def normalize_attr(entry_attrs, attr):
+    """
+    Convert all the entry attributes that match "attr" to same case as "attr"
+    """
+    vals = []
+    nattr = attr.lower()
+    for key_attr in entry_attrs:
+        if key_attr.lower() == nattr:
+            vals = entry_attrs[key_attr].copy()
+            del entry_attrs[key_attr]
+            break
+    if len(vals) > 0:
+        entry_attrs[attr] = vals
+
+
+def ensure_str(val):
+    """
+    Convert all non-binary values to strings for easier comparision
+    """
+    if val is not None and type(val) is not str:
+        try:
+            result = val.decode('utf-8')
+        except UnicodeDecodeError:
+            # binary value, must return it as is
+            result = val
+        return result
+    return val
+
+
+def ensure_list_str(val):
+    return [ensure_str(v) for v in val]
+
+
+def decode_attr_vals(entry_attrs):
+    """
+    Decode all attribute values for easier processing
+    """
+    decoded_attrs = {}
+    for attr in entry_attrs:
+        vals = ensure_list_str(entry_attrs[attr])
+        decoded_attrs[attr] = vals
+    return decoded_attrs
+
+
+def get_ldif_records(ldif_file, decode=True):
+    """
+    Returns a list of all the parsed records. Only run this on small LDIF
+    files as all the entries go into memory (cn=config/cn=schema is ok).
+    """
+    content = ldif.LDIFRecordList(open(ldif_file, "r"))
+    content.parse()
+
+    # LDIF entries look like [(dn, entry), (dn, entry), ...]
+    # dn = Str
+    # entry = {ATTR: [b'value', b'value', ...], ATTR: [], ...}
+    if decode:
+        entries = []
+        for dn, entry in content.all_records:
+            entries.append((dn, decode_attr_vals(entry)))
+        return entries
+    else:
+        # Return binary string as LDIFRecordList intended
+        return content.all_records
+
+
+def print_progress(msg):
+    sys.stdout.write('\r')
+    sys.stdout.write(msg)
+    sys.stdout.flush()
+
+
+#
+# ldif.LDIFParser requires that we create our own handler for parsing records
+# We need to set our class so we can call the IPAMigrate functions
+#
+class LDIFParser(ldif.LDIFParser):
+    mc = None
+    get_realm = False
+
+    def set_class(self, obj):
+        # Sets the IPAMigrate class
+        self.mc = obj
+
+    def look_for_realm(self):
+        self.get_realm = True
+
+    # Override handle() to do our specific migration work
+    def handle(self, dn, entry):
+        if self.mc is None:
+            return
+
+        if self.get_realm:
+            # Get the realmk from krb container
+            if DN("cn=kerberos," + str(self.mc.remote_suffix)) in DN(dn):
```

Comment:
Instead of + you can use , and the DN class will combine the strings. I'm curious what type the remote_suffix is though. If it is also a DN then you can use it without conversion.

Reply:
I think it is DN, I'll double check this...

Reply:
> I think it is DN, I'll double check this...

Yup, I was able to revise this, thanks!